File: frontend/src/lib/api/staff.ts
import { apiClient } from "@/lib/api/client";
import {
  staffResponseListSchema,
  staffResponseSchema,
  teacherOptionResponseListSchema,
} from "@/lib/schemas/staff";
import type { StaffCreate, StaffResponse, StaffType, StaffUpdate, TeacherOptionResponse } from "@/lib/types";
import { z } from "zod";

const payrollRateSchema = z.object({
  id: z.string().uuid(),
  rate_amount: z.number().int(),
  assignment_role: z.enum(["TEACHER", "ASSISTANT"]).nullable().nullish().default(null),
  effective_from: z.string(),
  effective_to: z.string().nullable(),
  version: z.number().int(),
});
const payrollSettlementSchema = z.object({
  id: z.string().uuid(), total_amount: z.number().int(), cutoff_at: z.string(),
  method: z.string(), reference: z.string().nullable(), created_at: z.string(),
  settlement_account_id: z.string().uuid().nullable().default(null),
  settlement_bank_code: z.string().nullable().default(null),
  settlement_bank_name: z.string().nullable().default(null),
  settlement_account_number: z.string().nullable().default(null),
  settlement_account_name: z.string().nullable().default(null),
  reversed_at: z.string().nullable(),
});
const payrollSettlementReversalSchema = z.object({
  id: z.string().uuid(), settlement_id: z.string().uuid(), staff_id: z.string().uuid(),
  request_id: z.string().uuid(), reason: z.string(), created_at: z.string(),
});
const payrollSummarySchema = z.object({
  staff_id: z.string().uuid(), balance: z.number().int(),
  rates: z.array(payrollRateSchema), settlements: z.array(payrollSettlementSchema),
});
export type StaffPayrollSummary = z.infer<typeof payrollSummarySchema>;

const attendanceHistoryItemSchema = z.object({
  attendance_id: z.string().uuid(),
  class_name: z.string().nullable(),
  role: z.string(),
  occurrence_start_at: z.string(),
  occurrence_end_at: z.string(),
  kind: z.string(),
  checkin_at: z.string(),
  rate_amount: z.number().int(),
  rate_version: z.number().int(),
  reversed_at: z.string().nullable().nullish().default(null),
  reversal_reason: z.string().nullable().nullish().default(null),
});
const attendanceHistorySchema = z.object({
  staff_id: z.string().uuid(),
  items: z.array(attendanceHistoryItemSchema),
});
export type StaffAttendanceHistory = z.infer<typeof attendanceHistorySchema>;

const manualAttendanceTargetSchema = z.object({
  occurrence_id: z.string().uuid(),
  class_name: z.string(),
  role: z.string(),
  occurrence_start_at: z.string(),
  occurrence_end_at: z.string(),
  kind: z.string(),
  rate_amount: z.number().int().nullable(),
});
export type ManualAttendanceTarget = z.infer<typeof manualAttendanceTargetSchema>;

const attendanceCheckInSchema = z.object({
  attendance_id: z.string().uuid(),
  status: z.literal("CHECKED_IN"),
  checkin_at: z.string(),
  rate_amount: z.number().int(),
  occurrence_start_at: z.string(),
});
export type AttendanceCheckIn = z.infer<typeof attendanceCheckInSchema>;

const attendanceReversalSchema = z.object({
  attendance_id: z.string().uuid(),
  reversed_at: z.string(),
  reason: z.string(),
});
export type AttendanceReversal = z.infer<typeof attendanceReversalSchema>;

type GetStaffParams = {
  staff_type?: StaffType;
  is_active?: boolean | null;
};

export async function getStaffMembers(params: GetStaffParams = {}): Promise<StaffResponse[]> {
  const { data } = await apiClient.get<StaffResponse[]>("/staff", {
    params: {
      staff_type: params.staff_type || undefined,
      is_active: params.is_active === null ? undefined : (params.is_active ?? true),
    },
  });
  return staffResponseListSchema.parse(data);
}

export async function getActiveStaffOptions(): Promise<TeacherOptionResponse[]> {
  const { data } = await apiClient.get<TeacherOptionResponse[]>("/staff/options");
  return teacherOptionResponseListSchema.parse(data);
}

export async function getActiveTeacherOptions(): Promise<TeacherOptionResponse[]> {
  return getActiveStaffOptions();
}

export async function createStaffMember(data: StaffCreate): Promise<StaffResponse> {
  const response = await apiClient.post<StaffResponse>("/staff", data);
  return staffResponseSchema.parse(response.data);
}

export async function updateStaffMember(
  id: string,
  data: StaffUpdate,
): Promise<StaffResponse> {
  const response = await apiClient.patch<StaffResponse>(`/staff/${id}`, data);
  return staffResponseSchema.parse(response.data);
}

export async function getStaffPayroll(id: string): Promise<StaffPayrollSummary> {
  const { data } = await apiClient.get(`/staff/${id}/payroll`);
  return payrollSummarySchema.parse(data);
}

export async function getStaffAttendanceHistory(
  id: string,
): Promise<StaffAttendanceHistory> {
  const { data } = await apiClient.get(`/staff/${id}/attendance-history`);
  return attendanceHistorySchema.parse(data);
}

export async function getManualAttendanceTargets(
  id: string,
): Promise<ManualAttendanceTarget[]> {
  const { data } = await apiClient.get(`/staff/${id}/attendance/manual-targets`);
  return z.array(manualAttendanceTargetSchema).parse(data);
}

export async function createManualAttendance(
  id: string,
  payload: { occurrence_id: string; request_id: string; reason?: string | null },
): Promise<AttendanceCheckIn> {
  const { data } = await apiClient.post(`/staff/${id}/attendance/manual`, payload);
  return attendanceCheckInSchema.parse(data);
}

export async function reverseAttendance(
  staffId: string,
  attendanceId: string,
  payload: { request_id: string; reason: string },
): Promise<AttendanceReversal> {
  const { data } = await apiClient.post(
    `/staff/${staffId}/attendance/${attendanceId}/reversal`,
    payload,
  );
  return attendanceReversalSchema.parse(data);
}

export async function createStaffCompensationRate(
  id: string,
  payload: {
    rate_amount: number;
    assignment_role?: StaffType | null;
    effective_from: string;
    effective_to?: string | null;
    reason?: string | null;
  },
) {
  const { data } = await apiClient.post(`/staff/${id}/compensation-rates`, payload);
  return payrollRateSchema.parse(data);
}

export async function settleStaffPayroll(
  id: string,
  payload: {
    request_id: string;
    method: "bank_transfer" | "cash";
    settlement_account_id?: string | null;
    reference?: string | null;
  },
) {
  const { data } = await apiClient.post(`/staff/${id}/payroll/settlements`, payload);
  return payrollSettlementSchema.parse(data);
}

export async function reverseStaffPayrollSettlement(
  staffId: string,
  settlementId: string,
  payload: { request_id: string; reason: string },
) {
  const { data } = await apiClient.post(
    `/staff/${staffId}/payroll/settlements/${settlementId}/reversal`,
    payload,
  );
  return payrollSettlementReversalSchema.parse(data);
}

File: backend/app/schemas/staff.py
from datetime import date, datetime
import re
from typing import Literal
from uuid import UUID

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    field_validator,
    model_validator,
)

from app.core.contact import validate_complete_contact_pair
from app.core.phone import is_valid_vietnam_mobile_phone, normalize_vietnam_phone

StaffType = Literal["TEACHER", "ASSISTANT"]


class StaffCompensationRateCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    rate_amount: int = Field(gt=0, le=999_999_999)
    assignment_role: StaffType | None = None
    effective_from: date
    effective_to: date | None = None
    reason: str | None = Field(default=None, max_length=500)

    @model_validator(mode="after")
    def validate_interval(self) -> "StaffCompensationRateCreate":
        if self.effective_to is not None and self.effective_to <= self.effective_from:
            raise ValueError("Ngày kết thúc hiệu lực phải sau ngày bắt đầu")
        return self


class StaffCompensationRateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    rate_amount: int
    assignment_role: StaffType | None = None
    effective_from: date
    effective_to: date | None
    version: int


class StaffPayrollSettlementCreate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    request_id: UUID
    method: Literal["bank_transfer", "cash"]
    settlement_account_id: UUID | None = None
    reference: str | None = Field(default=None, max_length=120)
    reason: str | None = Field(default=None, max_length=500)

    @model_validator(mode="after")
    def validate_settlement_account(self) -> "StaffPayrollSettlementCreate":
        if self.method == "bank_transfer" and self.settlement_account_id is None:
            raise ValueError("Hãy chọn tài khoản ngân hàng dùng để tất toán")
        if self.method == "cash" and self.settlement_account_id is not None:
            raise ValueError("Tất toán tiền mặt không cần tài khoản ngân hàng")
        return self


class StaffPayrollSettlementResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    total_amount: int
    cutoff_at: datetime
    method: str
    settlement_account_id: UUID | None
    settlement_bank_code: str | None
    settlement_bank_name: str | None
    settlement_account_number: str | None
    settlement_account_name: str | None
    reference: str | None
    created_at: datetime
    reversed_at: datetime | None = None


class StaffPayrollSettlementReversalCreate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    request_id: UUID
    reason: str = Field(min_length=1, max_length=500)


class StaffPayrollSettlementReversalResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    settlement_id: UUID
    staff_id: UUID
    request_id: UUID
    reason: str
    created_at: datetime


class StaffPayrollSummaryResponse(BaseModel):
    staff_id: UUID
    balance: int
    rates: list[StaffCompensationRateResponse]
    settlements: list[StaffPayrollSettlementResponse]


class StaffAttendanceHistoryItem(BaseModel):
    attendance_id: UUID
    class_name: str | None = None
    role: str
    occurrence_start_at: datetime
    occurrence_end_at: datetime
    kind: str
    checkin_at: datetime
    rate_amount: int
    rate_version: int
    reversed_at: datetime | None = None
    reversal_reason: str | None = None


class StaffAttendanceHistoryResponse(BaseModel):
    staff_id: UUID
    items: list[StaffAttendanceHistoryItem]


class StaffBase(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    full_name: str = Field(min_length=1, max_length=255)
    # Deprecated compatibility field. Roles are selected per class.
    staff_type: StaffType | None = None
    zalo_name: str | None = Field(default=None, max_length=100)
    phone: str | None = Field(default=None, max_length=32)
    email: str | None = Field(default=None, max_length=320)
    checkin_window_after_hours: int = Field(default=24, ge=1, le=720)
    is_active: bool = True

    @field_validator("zalo_name")
    @classmethod
    def normalize_zalo_name(cls, value: str | None) -> str | None:
        return value or None

    @field_validator("email")
    @classmethod
    def validate_email(cls, value: str | None) -> str | None:
        if not value:
            return None
        return _normalize_and_validate_email(value)

    @model_validator(mode="after")
    def validate_contact_pair(self) -> "StaffBase":
        validate_complete_contact_pair(
            zalo_name=self.zalo_name,
            phone=self.phone,
            owner="nhân sự",
        )
        return self


class StaffCreate(StaffBase):
    zalo_name: str = Field(min_length=1, max_length=100)
    phone: str = Field(min_length=1, max_length=32)

    @field_validator("phone")
    @classmethod
    def validate_phone(cls, value: str | None) -> str | None:
        return _normalize_and_validate_phone(value)


class StaffUpdate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    full_name: str | None = Field(default=None, min_length=1, max_length=255)
    staff_type: StaffType | None = None
    zalo_name: str | None = Field(default=None, max_length=100)
    phone: str | None = Field(default=None, max_length=32)
    email: str | None = Field(default=None, max_length=320)
    checkin_window_after_hours: int | None = Field(default=None, ge=1, le=720)
    is_active: bool | None = None

    @model_validator(mode="before")
    @classmethod
    def reject_null_for_required_columns(cls, value: object) -> object:
        if isinstance(value, dict):
            required_columns = {"full_name", "is_active"}
            null_fields = sorted(
                field
                for field in required_columns
                if field in value and value[field] is None
            )
            if null_fields:
                raise ValueError(f"Không được để trống: {', '.join(null_fields)}")
        return value

    @field_validator("phone")
    @classmethod
    def validate_phone(cls, value: str | None) -> str | None:
        return _normalize_and_validate_phone(value)

    @field_validator("zalo_name")
    @classmethod
    def normalize_zalo_name(cls, value: str | None) -> str | None:
        return value or None

    @field_validator("email")
    @classmethod
    def validate_email(cls, value: str | None) -> str | None:
        if not value:
            return None
        return _normalize_and_validate_email(value)


class StaffClassResponse(BaseModel):
    id: UUID
    name: str
    is_active: bool
    role: StaffType | None = None


class StaffResponse(StaffBase):
    id: UUID
    current_rate: int | None = None
    attendance_account_status: Literal[
        "connected", "disabled", "invited", "expired", "not_connected"
    ] = "not_connected"
    assigned_classes: list[StaffClassResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TeacherOptionResponse(BaseModel):
    id: UUID
    full_name: str
    staff_type: Literal["TEACHER", "ASSISTANT"] | None = None
    email: str | None = None


def _normalize_and_validate_phone(value: str | None) -> str | None:
    if (
        value is not None
        and value.strip()
        and not re.fullmatch(
            r"[0-9+().\s-]+",
            value,
        )
    ):
        raise ValueError("SĐT nhân sự chứa ký tự không hợp lệ")
    normalized = normalize_vietnam_phone(value)
    if normalized is None:
        return None
    if not is_valid_vietnam_mobile_phone(normalized):
        raise ValueError("SĐT nhân sự phải là số di động Việt Nam hợp lệ")
    return normalized


_EMAIL_PATTERN = re.compile(r"^[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}$")


def _normalize_and_validate_email(value: str | None) -> str | None:
    if value is None:
        return None
    normalized = value.strip().lower()
    if not normalized:
        return None
    if len(normalized) > 320:
        raise ValueError("Email nhân sự không được vượt quá 320 ký tự")
    if not _EMAIL_PATTERN.fullmatch(normalized):
        raise ValueError("Email nhân sự không hợp lệ")
    return normalized
